rust/src/passwd.rs

```diff
@@ -173,16 +174,152 @@ pub fn migrate_group_except_root(rootfs_dfd: i32, preserved_groups: &Vec<String>
     Ok(())
 }
 
-/// Merge and deduplicate entries from /usr/etc and /usr/lib.
-pub fn concat_fs_content(
+pub fn passwd_compose_prep(rootfs_dfd: i32, treefile: &mut Treefile) -> Result<()> {
+    let rootfs = ffiutil::ffi_view_openat_dir(rootfs_dfd);
+    passwd_compose_prep_impl(&rootfs, treefile, None, true)
+}
+
+/// Passwd/group handler for composes/treefiles.
+///
+/// We support various passwd/group handling. This function is primarily
+/// responsible for handling the "previous" and "file" paths; in both
+/// cases we inject data into the tree before even laying
+/// down any files, and notably before running RPM `useradd` etc.
+pub fn passwd_compose_prep_repo(
     rootfs_dfd: i32,
+    treefile: &mut Treefile,
     mut ffi_repo: Pin<&mut crate::ffi::OstreeRepo>,
     previous_checksum: &str,
+    unified_core: bool,
 ) -> Result<()> {
-    anyhow::ensure!(!previous_checksum.is_empty(), "missing previous reference");
-
     let rootfs = ffiutil::ffi_view_openat_dir(rootfs_dfd);
     let repo = ffi_repo.gobj_wrap();
+    passwd_compose_prep_impl(
+        &rootfs,
+        treefile,
+        Some((&repo, previous_checksum)),
+        unified_core,
+    )
+}
+
+fn passwd_compose_prep_impl(
+    rootfs: &openat::Dir,
+    treefile: &mut Treefile,
+    repo_previous_ref: Option<(&ostree::Repo, &str)>,
+    unified_core: bool,
+) -> Result<()> {
+    let generate_from_previous = treefile.parsed.preserve_passwd.unwrap_or(false);
+    if !generate_from_previous {
+        // Nothing to do
+        return Ok(());
+    };
+
+    let dest = if unified_core { "usr/etc" } else { "etc/" };
+
+    // Create /etc in the target root; FIXME - should ensure we're using
+    // the right permissions from the filesystem RPM.  Doing this right
+    // is really hard because filesystem depends on setup which installs
+    // the files...
+    rootfs.ensure_dir_all(dest, 0o0755)?;
+
+    let found_passwd_data = data_from_json(rootfs, treefile, dest, "passwd")?;
+    let found_groups_data = data_from_json(rootfs, treefile, dest, "group")?;
+
+    // We should error if we are getting passwd data from JSON and group from
+    // previous commit, or vice versa, as that'll confuse everyone when it goes
+    // wrong.
+    match (found_passwd_data, found_groups_data) {
+        (true, false) => {
+            anyhow::bail!("configured to migrate passwd data from JSON, and group data from commit")
+        }
+        (false, true) => {
+            anyhow::bail!("configured to migrate passwd data from commit, and group data from JSON")
+        }
+        _ => {}
+    };
+
+    if !found_passwd_data {
```

Comment:
@jlebon I think this is the correct translation of underlying logic introduced in https://github.com/coreos/rpm-ostree/pull/1704, but I'd appreciate a double-check.

Reply:
Yeah, that looks right to me! (Though minor bikeshed: `repo_previous_ref` -> `repo_previous_rev`/`repo_previous_checksum`.)